### PR TITLE
Add nodejs-function-invoker buildpack

### PR DESCRIPTION
**Pull request by edmorley**
This replaces the various Riff-based invoker buildpacks, now that Riff has been sunset. This buildpack's purpose is to vendor the new Node.js functions invoker:
https://github.com/forcedotcom/sf-fx-runtime-nodejs

For Riff invoker buildpack prior art, see:
https://github.com/forcedotcom/nodejs-sf-fx-buildpack
https://github.com/projectriff/node-function-buildpack

Also takes inspiration from the JVM invoker buildpack:
https://github.com/heroku/buildpacks-jvm/tree/main/buildpacks/jvm-function-invoker
...though with a number of changes since this repo has a different style of buildpack structure (to allow for unit testing), along with different buildpack utility functions (eg for logging), and uses shpec rather than Rapier for testing.

This initial version of the buildpack doesn't implement caching, since whilst the invoker is under active development, the release assets will be regularly changing, so using unversioned URLs and frequently changing archive SHA.

This buildpack also currently doesn't pass `$PORT` to the entrypoint, since the invoker currently doesn't support it (and may read directly from the env anyway, when there is support for custom ports).

The Circle CI shellcheck version has been bumped to pick up a fix that reduces the number of `source=` annotations required to prevent `SC1090` errors.

After this PR is merged and the buildpack published, the meta-buildpack `nodejs-function` will be updated in another PR (#48), to reference this buildpack instead of the Riff-based buildpacks.

Closes GUS-W-9115013.

### CODE REVIEW

**Comment by edmorley** [comment]
Local test using pack CLI and one of the fixtures (obviously doesn't do anything since not vendoring the real invoker, but at least shows detect/build works outside of the shpec tests):

```
$ pack build nodejs-fx-test --buildpack buildpacks/nodejs/ --buildpack buildpacks/nodejs-function-invoker/ --path buildpacks/nodejs-function-invoker/fixtures/function-toml/
20: Pulling from heroku/buildpacks
Digest: sha256:6ee998764c5ec6c63add69d6e133bba408d3d7af00d6f576d88d3b0e348cfe3d
Status: Image is up to date for heroku/buildpacks:20
20: Pulling from heroku/pack
Digest: sha256:e1ef17564fa514972373e63ceeb01d97a8b3e5cef79bee624b637b80a6fb61fc
Status: Image is up to date for heroku/pack:20
===> DETECTING
heroku/nodejs-engine           0.7.4
heroku/nodejs-function-invoker 0.1.0
===> ANALYZING
Previous image with name "nodejs-fx-test" not found
===> RESTORING
===> BUILDING
[INFO] Node.js Buildpack
[INFO] Setting NODE_ENV to production
[INFO] Installing toolbox
[INFO] - yj

[Installing Node]
[INFO] Getting Node version
[INFO] Resolving Node version
[INFO] Downloading and extracting Node v12.22.1

[Parsing package.json]
[INFO] Parsing package.json
[INFO] No file to start server
[INFO] either use 'docker run' to start container or add index.js or server.js

[Installing Node.js function runtime]
[INFO] Starting download of function runtime
[INFO] Function runtime installation successful
===> EXPORTING
Adding layer 'heroku/nodejs-engine:nodejs'
Adding layer 'heroku/nodejs-function-invoker:sf-fx-runtime-nodejs'
Adding 1/1 app layer(s)
Adding layer 'launcher'
Adding layer 'config'
Adding layer 'process-types'
Adding label 'io.buildpacks.lifecycle.metadata'
Adding label 'io.buildpacks.build.metadata'
Adding label 'io.buildpacks.project.metadata'
Setting default process type 'web'
Saving nodejs-fx-test...
*** Images (2eabc48c2edb):
      nodejs-fx-test
Adding cache layer 'heroku/nodejs-engine:nodejs'
Adding cache layer 'heroku/nodejs-engine:toolbox'
Successfully built image nodejs-fx-test
```

File: buildpacks/nodejs-function-invoker/buildpack.toml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+[[stacks]]
+id = "io.buildpacks.stacks.bionic"
+
+[metadata]
+
+[metadata.runtime]
+# TODO: Replace this with the actual invoker runtime archive URL.
+# This is currently using a random .tar.gz URL to test download/extraction.
+url = "https://nodejs.org/download/release/latest-v14.x/node-v14.16.1-linux-x64.tar.gz"
```
Comment (Malax) [inline buildpacks/nodejs-function-invoker/buildpack.toml:28]: I think we want to have some sort of validation of the downloaded file like the JVM function invoker buildpack does. We might deactivate it at first to allow quick changes in the invoker impl pre-beta, but having the feature implemented would be nice.

**Review by Malax — CHANGES_REQUESTED** [review]

File: buildpacks/nodejs-function-invoker/buildpack.toml (hunk 1 above)
Comment (edmorley) [reply buildpacks/nodejs-function-invoker/buildpack.toml:28]: I intentionally omitted that for now based on the conversation at:
https://gus.lightning.force.com/lightning/r/0D5AH0000011wd20AA/view

I'd kind of rather leave it out than have it half-commented out and not tested in this initial implementation?

Comment (edmorley) [reply buildpacks/nodejs-function-invoker/buildpack.toml:28]: We could add it in in later PRs fairly soon after this one - I was aiming for an MVP that we could iterate on afterwards to keep the PR small - though open to ideas :-)

Comment (Malax) [reply buildpacks/nodejs-function-invoker/buildpack.toml:28]: That works for me too! :)

**Review by Malax — APPROVED** [review]

**Review by Malax — APPROVED** [review]

File: buildpacks/nodejs-function-invoker/lib/build.sh [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+		launch = true
+		build = false
+		cache = false
+	EOF
+
+	runtime_tarball_url=$(toml_get_key_from_metadata "${CNB_BUILDPACK_DIR}/buildpack.toml" "runtime.url")
+
+	info "Starting download of function runtime"
+
+	if ! download_file "${runtime_tarball_url}" "${runtime_layer_dir}/sf-fx-runtime-nodejs.tar.gz"; then
```
Comment (edmorley) [inline buildpacks/nodejs-function-invoker/lib/build.sh:27]: @Malax I realised I forgot to mention this during our call yesterday - the runtime archive is downloaded into the layer directory and not deleted later, so will exist in the launch image. It seems we should either download to a temporary filepath, or else explicitly clean it up after the `npm install`. Thoughts?

Comment (edmorley) [reply buildpacks/nodejs-function-invoker/lib/build.sh:27]: I've filed W-9238585.

**Pull request merged**